### ISSUE TITLE
[rocprofiler-sdk] Inline queue interposition stalls at PyTorch profiler WAIT→WARMUP transition

### ISSUE THREAD

**Issue by lauri9** [ROCm/rocm-systems#11623]
## Problem Description

PyTorch Kineto profiling intermittently stalls GPU queues when the profiler transitions from `WAIT` to `WARMUP`. The first visible symptom is:

```text
queue_interposition.cpp:470] Async signal handler still waiting on signal ... (value=1, starting_value=1)
```

In a real 8-GPU Wan2.2 workload this left a DistVAE `ALLGATHER` unable to progress until its 600-second watchdog timeout. Setting `ROCPROFILER_QUEUE_INTERPOSITION=0` avoids both the warning and hang.

The Dockerfile below installs the latest compatible ROCm and PyTorch nightlies from TheRock's pip index. The minimal reproducer detects the exact warning and stops immediately. It reproduces in 5/5 trials; with `ROCPROFILER_QUEUE_INTERPOSITION=0`, it reproduces in 0/1 trials.

This appears related to the remaining 0→1 consumer race discussed during [#8212](https://github.com/ROCm/rocm-systems/pull/8212): queue resynchronization occurs before incrementing the consumer count.

## Operating System

Ubuntu 24.04.4 LTS

## CPU

AMD EPYC 9575F

## GPU

8 × AMD Instinct MI355X (`gfx950`)

## ROCm Version

Reproduced with TheRock nightlies published September 15, 2026:

- `rocm==10.2.0a20260915`
- `rocm-sdk-libraries==10.2.0a20260915`
- `torch==2.15.0a0+rocm10.2.0a20260915`
- HIP runtime: `7.16.26370`

## ROCm Component

rocprofiler-sdk

## Steps to Reproduce

Save the expandable Dockerfile and reproducer below as `docker/Dockerfile.qi-warmup-repro` and `scripts/qi_warmup_hang_repro_minimal.py`, then run:

```bash
docker build \
  -f docker/Dockerfile.qi-warmup-repro \
  -t qi-warmup-repro:nightly .

docker run --rm \
  --device=/dev/kfd --device=/dev/dri \
  --group-add video \
  --security-opt seccomp=unconfined \
  --ipc=host \
  qi-warmup-repro:nightly
```

Expected: profiling proceeds normally.

Actual:

```text
step 1/3 WAIT
step 2/3 WARMUP
USDT:2026-09-15 09:26:39 331:331 SyncActivityProfilerHandler.cpp:39] profiler_start
step 3/3 ACTIVE
W0915 09:26:50.263041 458 queue_interposition.cpp:470] Async signal handler still waiting on signal {.handle=139926395969152} after 1048576 iterations (value=1, starting_value=1)
RESULT trial=1 HIT elapsed=24.9s
SUMMARY hits=1/1
```

## Workaround

Setting `ROCPROFILER_QUEUE_INTERPOSITION=0` avoids the warning and the observed GPU/NCCL hang. It disables rocprofiler-sdk queue interposition, so users should confirm that the trace data they require remains available:

```bash
docker run --rm \
  --device=/dev/kfd --device=/dev/dri \
  --group-add video \
  --security-opt seccomp=unconfined \
  --ipc=host \
  -e ROCPROFILER_QUEUE_INTERPOSITION=0 \
  qi-warmup-repro:nightly \
  python /repro.py --timeout-s 30
```

This reports `SUMMARY hits=0/1`.

## Self-contained Reproducer

<details>
<summary><code>docker/Dockerfile.qi-warmup-repro</code></summary>

```dockerfile
FROM ubuntu:24.04

ARG ROCM_INDEX=https://nightly.repo.amd.com/rocm/whl-next/

RUN apt-get update \
    && apt-get install -y --no-install-recommends \
        ca-certificates \
        libgomp1 \
        libnuma1 \
        python3 \
        python3-venv \
    && rm -rf /var/lib/apt/lists/*

RUN python3 -m venv /opt/venv \
    && /opt/venv/bin/pip install --upgrade pip numpy \
    && /opt/venv/bin/pip install --pre --index-url "${ROCM_INDEX}" \
        "torch[device-gfx950]"

COPY scripts/qi_warmup_hang_repro_minimal.py /repro.py

ENV PATH=/opt/venv/bin:${PATH}

CMD ["python", "/repro.py", "--trials", "5"]
```

</details>

<details>
<summary><code>scripts/qi_warmup_hang_repro_minimal.py</code></summary>

```python
#!/usr/bin/env python3
"""Fast reproducer for the rocprofiler-sdk queue-interposition WARMUP race.

Run inside the affected image:

    python qi_warmup_hang_repro_minimal.py

The launcher exits successfully on the first exact SDK warning. Use
``--trials 5`` to check reliability. Eight GPUs are required.
"""

from __future__ import annotations

import argparse
import datetime
import os
import selectors
import signal
import subprocess
import sys
import threading
import time

import torch
from torch.profiler import ProfilerActivity, profile, schedule

WARNING_PARTS = ("queue_interposition.cpp", "value=1")
WORLD_SIZE = 8
STREAMS = 16
MATRIX_SIZE = 2048
LAUNCHES = 128
COLLECTIVES = 32


def run_trials(trials: int, timeout_s: float) -> int:
    hits = 0
    for trial in range(1, trials + 1):
        command = [
            sys.executable,
            "-m",
            "torch.distributed.run",
            f"--nproc_per_node={WORLD_SIZE}",
            os.path.abspath(__file__),
            "--worker",
        ]
        process = subprocess.Popen(
            command,
            stdout=subprocess.PIPE,
            stderr=subprocess.STDOUT,
            text=True,
            bufsize=1,
            start_new_session=True,
        )
        assert process.stdout is not None
        selector = selectors.DefaultSelector()
        selector.register(process.stdout, selectors.EVENT_READ)
        started = time.monotonic()
        hit = False

        while process.poll() is None and time.monotonic() - started < timeout_s:
            if not selector.select(timeout=0.2):
                continue
            line = process.stdout.readline()
            if not line:
                continue
            print(line, end="", flush=True)
            if all(part in line for part in WARNING_PARTS):
                hit = True
                break

        if process.poll() is None:
            os.killpg(process.pid, signal.SIGTERM)
            try:
                process.wait(timeout=10)
            except subprocess.TimeoutExpired:
                os.killpg(process.pid, signal.SIGKILL)
                process.wait()
        selector.close()

        elapsed = time.monotonic() - started
        hits += int(hit)
        print(f"RESULT trial={trial} {'HIT' if hit else 'MISS'} elapsed={elapsed:.1f}s")

    print(f"SUMMARY hits={hits}/{trials}")
    return 0 if hits == trials else 1


def gemm(x: torch.Tensor) -> None:
    x.copy_(torch.mm(x, x))


def step_work(
    streams: list[torch.cuda.Stream],
    matrices: list[torch.Tensor],
    gather_input: torch.Tensor,
    gather_output: list[torch.Tensor],
) -> None:
    for index in range(LAUNCHES):
        with torch.cuda.stream(streams[index % STREAMS]):
            gemm(matrices[index % STREAMS])
    for _ in range(COLLECTIVES):
        torch.distributed.all_gather(gather_output, gather_input)
    torch.cuda.synchronize()


def storm(streams: list[torch.cuda.Stream], matrices: list[torch.Tensor]) -> None:
    index = 0
    while True:
        with torch.cuda.stream(streams[index % STREAMS]):
            gemm(matrices[index % STREAMS])
        index += 1


def run_worker() -> int:
    rank = int(os.environ["RANK"])
    torch.cuda.set_device(int(os.environ["LOCAL_RANK"]))
    torch.distributed.init_process_group(
        backend="nccl",
        timeout=datetime.timedelta(seconds=180),
    )

    streams = [torch.cuda.Stream() for _ in range(STREAMS)]
    matrices = [
        torch.randn(MATRIX_SIZE, MATRIX_SIZE, device="cuda")
        for _ in range(STREAMS)
    ]
    gather_input = torch.ones(8, device="cuda")
    gather_output = [
        torch.empty_like(gather_input)
        for _ in range(torch.distributed.get_world_size())
    ]
    work = lambda: step_work(streams, matrices, gather_input, gather_output)

    for _ in range(3):
        work()
    torch.distributed.barrier()
    threading.Thread(target=storm, args=(streams, matrices), daemon=True).start()

    with profile(
        activities=[ProfilerActivity.CPU, ProfilerActivity.CUDA],
        schedule=schedule(wait=1, warmup=1, active=1),
        record_shapes=True,
    ) as profiler:
        for step, phase in enumerate(("WAIT", "WARMUP", "ACTIVE"), start=1):
            if rank == 0:
                print(f"step {step}/3 {phase}", flush=True)
            work()
            profiler.step()
    return 0


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--trials", type=int, default=1)
    parser.add_argument("--timeout-s", type=float, default=45.0)
    parser.add_argument("--worker", action="store_true", help=argparse.SUPPRESS)
    args = parser.parse_args()
    return run_worker() if args.worker or "RANK" in os.environ else run_trials(args.trials, args.timeout_s)


if __name__ == "__main__":
    sys.exit(main())
```

</details>

## Related Issues and Changes

- [vLLM #54087 — torch profiler hangs after rocprofiler-sdk 1.3.2 bump](https://github.com/vllm-project/vllm/issues/54087)
- [#7796 — prevent queue-interposition doorbell over-advance](https://github.com/ROCm/rocm-systems/pull/7796)
- [#7924 — fix queue-interposition completion waiter ordering](https://github.com/ROCm/rocm-systems/pull/7924)
- [#8093 — bypass idle inline queue interposition without active consumers](https://github.com/ROCm/rocm-systems/pull/8093)
- [#8212 — backport of idle-bypass change; review discusses the remaining race](https://github.com/ROCm/rocm-systems/pull/8212)
- [vLLM #56521 — separate failure exposed when queue interposition is disabled](https://github.com/vllm-project/vllm/issues/56521)

### Operating System

Ubuntu 24.04.4 LTS

### CPU

AMD EPYC 9575F

### GPU

8xMI355X

### ROCm Version

ROCm 10.2 nightly 2029-09-15

### Installation Method

Python wheels

### Installed ROCm Packages / Versions

<details>
<summary>Installed ROCm packages / versions</summary>

```
- `rocm==10.2.0a20260915`
- `rocm-sdk-libraries==10.2.0a20260915`
- `torch==2.15.0a0+rocm10.2.0a20260915`
- HIP runtime: `7.16.26370`
```

</details>

### ROCm Component

rocprofiler-sdk

### (Optional for Linux users) Output of rocminfo --support

<details>
<summary>rocminfo --support output</summary>

```
Paste output here
```

</details>

**Comment by lauri9** [comment]
This no longer reproduces on the latest nightly build rocm 10.2.0a20260918.

Possibly https://github.com/ROCm/rocm-systems/pull/9984 has fixed it - closing
